Marlin/src/HAL/HAL_STM32/fastio_STM32.h

```diff
@@ -24,28 +24,55 @@
 
 /**
  * Fast I/O interfaces for STM32
- * These use GPIO functions instead of Direct Port Manipulation, as on AVR.
+ * These use GPIO register access for fast port manipulation.
  */
 
+// --------------------------------------------------------------------------
+// Public Variables
+// --------------------------------------------------------------------------
+
+extern GPIO_TypeDef * FastIOPortMap[];
+
+// --------------------------------------------------------------------------
+// Public functions
+// --------------------------------------------------------------------------
+
+void FastIO_init(); // Must be called before using fast io macros
+
+// --------------------------------------------------------------------------
+// Defines
+// --------------------------------------------------------------------------
+
 #define _BV(b) (1 << (b))
 
-#define READ(IO)                digitalRead(IO)
-#define WRITE(IO,V)             digitalWrite(IO,V)
-#define WRITE_VAR(IO,V)         WRITE(IO,V)
+#if defined(STM32F0xx) || defined(STM32F1xx) || defined(STM32F3xx) || defined(STM32L0xx) || defined(STM32L4xx)
+  #define _WRITE(IO, V) do { \
+    if (V) FastIOPortMap[STM_PORT(digitalPin[IO])]->BSRR = (1UL << STM_PIN(digitalPin[IO])) ; \
+    else FastIOPortMap[STM_PORT(digitalPin[IO])]->BRR = (1UL << STM_PIN(digitalPin[IO])) ; \
+  } while(0)
+#else
+  #define _WRITE(IO, V) (FastIOPortMap[STM_PORT(digitalPin[IO])]->BSRR = (1UL << STM_PIN(digitalPin[IO]) << (V ? 0UL : 16UL)))
+#endif
+
+#define _READ(IO)               bool(READ_BIT(FastIOPortMap[STM_PORT(digitalPin[IO])]->IDR, 1UL << STM_PIN(digitalPin[IO])))
+#define _TOGGLE(IO)             (FastIOPortMap[STM_PORT(digitalPin[IO])]->ODR ^= (1UL << STM_PIN(digitalPin[IO])))
 
 #define _GET_MODE(IO)
 #define _SET_MODE(IO,M)         pinMode(IO, M)
 #define _SET_OUTPUT(IO)         pinMode(IO, OUTPUT)                               /*!< Output Push Pull Mode & GPIO_NOPULL   */
 
+#define WRITE_VAR(IO,V)         _WRITE(IO,V)
+#define WRITE(IO,V)             _WRITE(IO,V)
+#define READ(IO)                _READ(IO)
+#define TOGGLE(IO)              _TOGGLE(IO)
+
 #define OUT_WRITE(IO,V)         do{ _SET_OUTPUT(IO); WRITE(IO,V); }while(0)
 
 #define SET_INPUT(IO)           _SET_MODE(IO, INPUT)                              /*!< Input Floating Mode                   */
 #define SET_INPUT_PULLUP(IO)    _SET_MODE(IO, INPUT_PULLUP)                       /*!< Input with Pull-up activation         */
 #define SET_INPUT_PULLDOWN(IO)  _SET_MODE(IO, INPUT_PULLDOWN)                     /*!< Input with Pull-down activation       */
 #define SET_OUTPUT(IO)          OUT_WRITE(IO, LOW)
 
```

Comment:
> SET_OUTPUT(IO) ... Is it a requirement to also set the pin LOW?

I'd say yes. A defined state is required. If you don't want a transition from input to LOW you could use OUT_WRITE(IO,HIGH).

Thanks.